File: wslpath.php
#!/usr/bin/php
<?php

// Auto-detects path - converts to a Unix path if it is a Windows path and vice-versa
// Correctly handles symlinks
// Correctly handles paths under lxss directory (hidden user directory under which Unix-only files and directories are located).
// Correctly handles :line:column suffix. For example, `subl $(wslpath /mnt/d/script.js:10:2)` would open D:\script.js at line 10, column 2 in Sublime Text

function usage() {
return 'wslpath [-m|-u|-w|-h] NAME[:line[:col]]

Output type options:

  -w           (default) prints Windows form of NAME (C:\WINNT)
  -m           like -w, but with regular slashes (C:/WINNT)
  -u           prints Unix form of NAME (/mnt/c/winnt)

Other options:

  -h           displays usage information

If no output type is selected, the program will try to detect the form of
NAME and print the opposite type (eg. will print Windows form for Unix
path).';
}

$lxssPath_ = null;
function lxssPath() {
	global $lxssPath_;
	if ($lxssPath_) return $lxssPath_;
	$lxssPath_ = trim(exec('cmd.exe /c "echo %LOCALAPPDATA%" 2> /dev/null')) . "\\lxss";
	return $lxssPath_;
}

function pathType($path) {
	preg_match('/^([A-Za-z]):(.*)$/', $path, $r);
	if (count($r) > 1) return 'win';
	if (strpos($path, '/') === 0) return 'unix';
	if (strpos($path, '~') === 0) return 'unix';
	if (strpos($path, "\\") !== false) return 'win';
	if (strpos($path, '/') !== false) return 'unix';
	return 'unix';
}

function slashJoin($part1, $part2, $slash) {
	$part1 = rtrim($part1, "/\\");
	$part2 = ltrim($part2, "/\\");
	return $part1 . $slash . $part2;
}

function setLastSlash($path, $addIt, $slash) {
	$path = rtrim($path, "/\\");
	if (!$addIt) return $path;
	return $path . $slash;
}

function toSlash($path, $slash) {
	$path = str_replace("\\", $slash, $path);
	$path = str_replace('/', $slash, $path);
	return $path;
}

function explodeLineColNumbers($path) {
	preg_match('/(.*):(\d+):(\d+)/', $path, $r);
	if (count($r) >= 4) {
		return array(
			'path' => $r[1],
			'line' => $r[2],
			'col' => $r[3],
		);
	}

	preg_match('/(.*):(\d+)/', $path, $r);
	if (count($r) >= 3) {
		return array(
			'path' => $r[1],
			'line' => $r[2],
		);
	}

	return array('path' => $path);
}

function pathToWin($inputPath) {
	$path = $inputPath[0] != '/' ? slashJoin(getcwd(), $inputPath, '/') : $inputPath;
	$path = realpath($path);
	if (!$path) return toSlash($inputPath, "\\");

	if (strpos($path, '/mnt/') === 0) {
		$initialDir = getcwd();

		$inputDir = dirname($path);

		$ok = @chdir($inputDir);
		if (!$ok) return toSlash($inputPath, "\\");

		$output = trim(exec('cmd.exe /c cd 2> /dev/null', $output));
		$output = slashJoin($output, basename($path), "\\");
	} else {
		$output = slashJoin(lxssPath(), $path, "\\");
	}

	$c = $inputPath[strlen($inputPath) - 1];
	$output = setLastSlash($output, $c == '/' || $c == "\\", "\\");

	return toSlash($output, "\\");
}

function pathToUnix($winPath) {
	preg_match('/([A-Za-z]):(.*)/', $winPath, $r);
	
	if (count($r) < 3) {
		$output = slashJoin(getcwd(), $winPath, '/');
	} else {
		$drive = '/mnt/' . strtolower($r[1]);
		$output = slashJoin($drive, $r[2], '/');
	}

	$c = $winPath[strlen($winPath) - 1];
	$output = setLastSlash($output, $c == '/' || $c == "\\", "\\");

	return toSlash($output, '/');
}

function main($args) {
	array_splice($args, 0, 1);

	$operation = null;
	$inputPath = null;
	$mixedMode = false;

	while (count($args)) {
		$arg = $args[0];

		if ($arg == '-h') {
			echo usage() . "\n";
			return;
		}

		if ($arg == '-w') {
			$operation = 'to_win';
			array_splice($args, 0, 1);
			continue;
		}

		if ($arg == '-u') {
			$operation = 'to_unix';
			array_splice($args, 0, 1);
			continue;
		}

		if ($arg == '-m') {
			$mixedMode = true;
			$operation = 'to_win';
			array_splice($args, 0, 1);
			continue;
		}

		if (strpos($arg, '-') !== 0) {
			$inputPath = trim($arg);
			break;
		}

		throw new Exception('Unknown option: ' . $arg);
	}

	if (!$inputPath) throw new Exception('No path provided');

	if (!$operation) {
		$operation = pathType($inputPath) == 'win' ? 'to_unix' : 'to_win';
	}

	$pathInfo = explodeLineColNumbers($inputPath);
	$outputPath = '';

	if ($operation == 'to_win') {
		$outputPath = pathToWin($pathInfo['path']);
		if ($mixedMode) $outputPath = toSlash($outputPath, '/');
	}

	if ($operation == 'to_unix') {
		$outputPath = pathToUnix($pathInfo['path']);
	}

	if (isset($pathInfo['line'])) $outputPath .= ':' . $pathInfo['line'];
	if (isset($pathInfo['col'])) $outputPath .= ':' . $pathInfo['col'];

	echo $outputPath . "\n";
}

try {
	main($argv);
} catch (Exception $e) {
	echo $e->getMessage() . "\n\n" . usage() . "\n";
	die(1);
}

File: wslpath.py
#! /usr/bin/env python3

from os import getcwd
from os import chdir
from os.path import basename
from sys import argv
import re

def dirname(path: str) -> str:
    return basename(path)

def usage() -> str:
    return '''wslpath [-m|-u|-w|-h] NAME[:line[:col]]

    Output type options:

    \t-w\t\t(default) prints Windows form of NAME (C:\WINNT)
    \t-m\t\tlike -w, but with regular slashes (C:/WINNT)
    \t-u\t\tprints Unix form of NAME (/mnt/c/winnt)

    Other options:

    \t-h\t\tdisplays usage information

    If no output type is selected, the program will try to detect the form of
    NAME and print the opposite type (eg. will print Windows form for Unix
    path).'''


lxssPath_ = "/mnt/c"
def lxssPath(path: str =None) -> str:
    global lxssPath_
    if path:
        lxssPath_ = path
        # lxssPath_ = trim(exec('cmd.exe /c "echo %LOCALAPPDATA%" 2> /dev/null')) . "\\lxss"
    return lxssPath_


def pathType(path) -> str:
    r = re.match('/^([A-Za-z]):(.*)$/', path)
    if r:
        return "win"
    if path[0] == "/" or path[0] == "~":
        return "unix"
    if "\\" in path:
        return "win"
    if "/" in path:
        return "unix"
    return "unix"


def slashJoin(part1: str, part2: str, slash: str) -> str:
    part1 = part1.rstrip()
    part1 = part1.removesuffix("/\\")
    part2 = part2.lstrip()
    part2 = part2.removeprefix("/\\")
    return part1 + slash + part2


def setLastSlash(path: str, addIt: bool, slash: str) -> str:
    path = path.rstrip().removesuffix("/\\")
    if (not addIt):
        return path
    return path + slash


def toSlash(path: str, slash: str) -> str:
    path = path.replace("\\", slash)
    path = path.replace("/", slash)
    return path


def explodeLineColNumbers(path: str) -> dict:
    r = re.match('/(.*):(\d+):(\d+)/', path)
    if r.end() >= 4:
        return {
            'path': r[1],
            'line': r[2],
            'col': r[3],
        }
    

    r = re.match('/(.*):(\d+)/', path)
    if r.end() >= 3:
        return {
            'path': r[1],
            'line': r[2],
        }
    

    return {
            'path': r[1],
        }


def pathToWin(inputPath: str) -> str:
    path = slashJoin(getcwd(), inputPath, '/') if inputPath[0] != '/' else inputPath
    if (not path):
        return toSlash(inputPath, "\\")

    if path.startswith("/mnt/"):
        initialDir = getcwd()

        inputDir = dirname(path)

        try:
            chdir(inputDir)
        except:
            return toSlash(inputPath, "\\")

        # output = trim(exec('cmd.exe /c cd 2> /dev/null', output))
        output = slashJoin(getcwd(), basename(path), "\\")
    else:
        output = slashJoin(lxssPath(), path, "\\")
    

    c = inputPath[len(inputPath) - 1]
    output = setLastSlash(output, c == '/' or c == "\\", "\\")

    return toSlash(output, "\\")


def pathToUnix(winPath: str) -> str:
    r = re.match('/([A-Za-z]):(.*)/', winPath)
    
    if r.end() < 3:
        output = slashJoin(getcwd(), winPath, '/')
    else:
        drive = '/mnt/' + r[1].lower()
        output = slashJoin(drive, r[2], '/')
    

    c = winPath[len(winPath) - 1]
    output = setLastSlash(output, c == '/' or c == "\\", "\\")

    return toSlash(output, '/')


def main(args):
    args = args[1:]

    operation = None
    inputPath = None
    mixedMode = False

    for i, arg in enumerate(args):

        try:
            if arg == '-h':
                print(usage())
                return

            if arg == '-w':
                operation = 'to_win'
                continue
            

            if arg == '-u':
                operation = 'to_unix'
                continue
            

            if arg == '-m':
                mixedMode = True
                operation = 'to_win'
                continue
            

            if "-" in arg and arg.find("-") != i:
                inputPath = arg.strip()
                break
        except:
            raise Exception('Unknown option: ' + arg)
    

    if not inputPath:
        raise Exception('No path provided')

    if not operation:
        operation = 'to_unix' if pathType(inputPath) == 'win' else 'to_win'
    

    pathInfo = explodeLineColNumbers(inputPath)
    outputPath = ''

    if operation == 'to_win':
        outputPath = pathToWin(pathInfo['path'])
        if mixedMode:
            outputPath = toSlash(outputPath, '/')
    

    if operation == 'to_unix':
        outputPath = pathToUnix(pathInfo['path'])
    

    if pathInfo['line']:
        outputPath += ':' + pathInfo['line']
    if pathInfo['col']:
        outputPath += ':' + pathInfo['col']

    print(outputPath)


if __name__ == "__main":
    try:
        main(argv)
    except Exception as e:
        print(str(e.args) + "\n\n" + usage() + "\n")
        exit()
